### PR TITLE
planner/cascades: add rule InjectProjectionBelowSort

### LINKED ISSUE #15061: planner/cascades: implement the transformation rule `InjectProjectionBelowSort`
## Feature Request

**Is your feature request related to a problem? Please describe:**

In TiDB's executor, `Sort` does not support scalar functions. But if we have the SQL like this: `select a from t order by a+b`, TiDB will uses a `Projection` to calculate the expression `a+b`, then use a `Sort` to sort the rows. 

#14504 has implemented a rule `InjectProjectionBelowTopN`. We also need a rule like this for `LogicalSort`. And then we can remove this line: https://github.com/pingcap/tidb/blob/master/planner/implementation/sort.go#L47.

Notice that this rule should be inside the `PostTransformationBatch`.

**Describe the feature you'd like:**

Implement the transformation rule `InjectProjectionBelowSort` with the same logic of `InjectProjectionBelowTopN`.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #15061

Problem Summary:

### What is changed and how it works?

What's Changed:

Add a new rule `InjectProjectionBelowSort` to `PostTransformationBatch` of cascades optimizer.

How it Works:

As explained in issue #15061, TiDB's executor doesn't support `Sort` containing scalar functions. So we need to inject a `Projection` below `Sort` to compute these scalar functions in advance. Then add a `Projection` above `Sort` to restore the original schema.
This is almost the same as `InjectProjectionBelowTopN`.

### Check List

Tests 

- Unit test

### Release note

- No release note

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -2222,6 +2225,102 @@ func (r *InjectProjectionBelowTopN) OnTransform(old *memo.ExprIter) (newExprs []
...
+	baseRule
+}
+
+// NewRuleInjectProjectionBelowSort creates a new Transformation NewRuleInjectProjectionBelowSort.
+// The behavior and reason of this rule is almost the same as `InjectProjectionBelowTopN` above.
+// It will extract the ScalarFunctions of `ByItems` into a Projection and injects the Projection
+// below Sort. Then, we need to add another Projection upon Sort to prune the extra Columns.
+// The reason why we need this rule is that, SortExecutor in TiDB does not support ScalarFunction
+// in `ByItems`. So we have to use a Projection to calculate the ScalarFunctions in advance.
+// The pattern of this rule is: a single Sort
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go:2240]:
```suggestion
// The pattern of this rule is: a single Sort.
```

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -2222,6 +2225,102 @@ func (r *InjectProjectionBelowTopN) OnTransform(old *memo.ExprIter) (newExprs []
...
+		if _, ok := item.Expr.(*expression.ScalarFunction); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// OnTransform implements Transformation interface.
+// It will convert `Sort -> X` to `Projection -> Sort -> Projection -> X`.
+func (r *InjectProjectionBelowSort) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go:2263]: Could we reuse the codes in `InjectProjectionBelowTopN`? I think they have the same logic.

Comment (time-and-fate) [reply planner/cascades/transformation_rules.go:2274]: Seems like we can't use only one rule for two different patterns. So I tried to use a new function `generateProj4Inject` to handle some of the same logic. Is this OK?
And sorry for being late.

Comment (francis0407) [reply planner/cascades/transformation_rules.go:2274]: Yeah, I'm OK with this solution.

**Comment by francis0407** [comment]
PTAL @winoros @lzmhhh123

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Pull request closed without merging**